c10/core/MobileAllocator.cpp

```diff
@@ -0,0 +1,106 @@
+#include <c10/core/MobileAllocator.h>
+#include <c10/core/CPUAllocator.h>
+
+namespace c10 {
+namespace {
+
+struct Configuration final {
+#if defined(__arm__) || defined(__aarch64__) || defined(_M_ARM) || defined(_M_ARM64)
+  static constexpr uint32_t kAlignment = 16u;
+#else
+  static constexpr uint32_t kAlignment = gAlignment;
+#endif
+};
+
+// QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors.
+// This behavior will trigger ASAN, and may result in a segfault if the accessed
+// memory location just so happens to fall on a page the current process has no
+// read access to.  Here we define a custom allocator that allocates the extra
+// storage required to keep this behavior safe.  This allocator could have been
+// restricted to QNNPACK and XNNPACK only, but that would have negative
+// performance ramifications, as input tensors must be reallocated, and copied
+// over, if not allocated with this allocator to begin with.  Making this
+// allocator the default on mobile builds minimizes the probability of unnecessary
+// reallocations and copies.
+//
+//
+// PreGuardBytes: Number of guard bytes to allocate before the allocation.
+// PostGuardBytes: Number of guard bytes to allocate after the allocation.
+
+template <uint32_t PreGuardBytes, uint32_t PostGuardBytes>
+class DefaultMobileCPUAllocator final : public at::Allocator {
+ public:
+  DefaultMobileCPUAllocator() = default;
+  virtual ~DefaultMobileCPUAllocator() override = default;
+
+  static void deleter(void* const pointer) {
+    if (C10_UNLIKELY(!pointer)) {
+      return;
+    }
+
+    const Cast memory{pointer};
+    c10::free_cpu(memory.as_byte_ptr - PreGuardBytes);
```

Comment:
It looks like `pointer` here is the context pointer, rather than the data pointer.  We could set that to the original allocated memory (not the offset value) and avoid the arithmetic here.

Reply:
Sorry I'm not following ... :/

Reply:
In allocate, you construct a DataPtr with two fields, the pointer to the data, and the "context" pointer, which is what gets passed to free here.  You set both to the "allocated pointer plus pre-guard bytes" value.  But if you set the second one to the original allocated value, you could just pass that to free_cpu here.

Reply:
I see, nice!

Reply:
Done.